src/Data/Array.purs

```diff
@@ -1118,13 +1146,82 @@ deleteBy eq x ys = maybe ys (\i -> unsafePartial $ fromJust (deleteAt i ys)) (fi
+difference :: forall a. Ord a => Array a -> Array a -> Array a
+difference = differenceBy compare
 
 infix 5 difference as \\
 
+differenceBy :: forall a. (a -> a -> Ordering) -> Array a -> Array a -> Array a
+differenceBy   _ left       [] = left
+differenceBy   _ left@[]     _ = left
+differenceBy cmp left    right = map snd $ ST.run do
+  result <- STA.new
+  latestRightArrayValue <- STRef.new Nothing
+  ST.foreach indexedAndSorted \(Tuple fromLeftArray pair@(Tuple _ x')) -> do
+    maybeValueToRemove <- STRef.read latestRightArrayValue
+    case fromLeftArray, maybeValueToRemove of
+      true, Just (Tuple valueToRemove count) | cmp valueToRemove x' == EQ ->
+        -- do not add left array's element to final array; check count
+        if count == 1 then do
+          void $ STRef.write Nothing latestRightArrayValue
+        else do
+          let decrementCount = Just (Tuple valueToRemove (count - 1))
+          void $ STRef.write decrementCount latestRightArrayValue
+
+      true, _ -> void $ STA.push pair result
+
+      false, Just (Tuple valueToRemove count) | cmp valueToRemove x' == EQ -> do
+        let next = if count == 0 then Nothing
+                   else Just (Tuple valueToRemove (count - 1))
+        void $ STRef.write next latestRightArrayValue
+      false, _ -> do
+        void $ STRef.write (Just (Tuple x' 1)) latestRightArrayValue
+  _ <- STA.sortWith fst result
+  STA.unsafeFreeze result
+  where
+    -- Note: when elements are equal, right array elements
+    -- (i.e. `Tuple false (Tuple _ _)`) appear "before" left array elements
+    -- (i.e. `Tuple true (Tuple _ _)`) in the resulting array.
+    valueThenRightFirst :: Tuple Boolean (Tuple Int a) -> Tuple Boolean (Tuple Int a) -> Ordering
+    valueThenRightFirst (Tuple lb (Tuple _ lv)) (Tuple rb (Tuple _ rv)) =
+      cmp lv rv <> compare lb rb
+
+    indexedAndSorted = sortBy valueThenRightFirst $ combineIndex left right
+
+-- Internal use only
+-- Essentially...
+-- ```
+-- combineWithIndex cmp left right =
+-- let
+--   leftIndexed = mapWithIndex (\idx v -> Tuple idx v) left
+--
+--   tupleAdjusted idx v = Tuple (idx + (length left)) v
+--   rightIndexedPlus = mapWithIndex tupleAdjusted right
+--
+-- in leftIndexed <> rightIndexedPlus
+-- ```
+-- ... but without creating two intermediate arrays due to the `mapWithIndex`
+-- on both arrays.
+-- ```
+-- let t3 a b c = Tuple a (Tuple b c)
+-- combineIndex compare [0] [0, 1]
+--   == [t3 true 0 0, t3 false 1 0, t3 false 2 1]
+-- ```
+combineIndex :: forall a. Array a -> Array a -> Array (Tuple Boolean (Tuple Int a))
+combineIndex left right = ST.run do
+  out <- STA.new
+  let leftLen = length left
+  ST.for 0 leftLen \idx -> do
+    let val = unsafePartial $ unsafeIndex left idx
+    void $ STA.push (Tuple true (Tuple idx val)) out
+  let rightLen = length right
+  ST.for 0 rightLen \idx -> do
+    let val = unsafePartial $ unsafeIndex right idx
+    void $ STA.push (Tuple false (Tuple (leftLen + idx) val)) out
```

Comment:
Are there any performance penalties to using a record instead of a tuple? Luckily, this tuple happens to be type-safe because of the distinct types (`Boolean`, `Int`, `a`).

Reply:
Not sure. I also tried it out with `data Tuple3 a b c = Tuple3 a b c` and that sped things up only slightly.